lib/galaxy/celery/base_task.py

```diff
@@ -0,0 +1,163 @@
+import datetime
+from abc import abstractmethod
+
+from celery import Task
+from sqlalchemy import (
+    bindparam,
+    insert,
+    select,
+    text,
+    update,
+)
+from sqlalchemy.dialects.postgresql import insert as ps_insert
+from sqlalchemy.exc import IntegrityError
+from sqlalchemy.orm import Session
+
+from galaxy.model import CeleryUserRateLimit
+from galaxy.model.base import transaction
+from galaxy.model.scoped_session import galaxy_scoped_session
+
+
+class GalaxyTaskBeforeStart:
+    """
+    Class used by the custom celery task, GalaxyTask, to implement
+    logic to limit number of task executions per user per second.
+    This superclass is used directly when no user rate limit logic
+    is to be enforced based on value of config param,
+    celery_user_rate_limit.
+    """
+
+    def __call__(self, task: Task, task_id, args, kwargs):
+        pass
+
+
+class GalaxyTaskBeforeStartUserRateLimit(GalaxyTaskBeforeStart):
+    """
+    Used when we wish to enforce a user rate limit based on
+    non-default value of celery_user_rate_limit config setting.
+    We limit executions by keeping track in a table of the last scheduled
+    time for the execution of a task by user. When a new task
+    is to be executed we schedule it a certain time interval
+    after the last scheduled execution of a task by this user
+    by doing a task.retry.
+    If the last scheduled execution was far enough in the past
+    then we allow the task to run immediately.
+    """
+
+    def __init__(
+        self,
+        tasks_per_user_per_sec: float,
+        ga_scoped_session: galaxy_scoped_session,
+    ):
+        self.task_exec_countdown_secs = 1 / tasks_per_user_per_sec
+        self.ga_scoped_session = ga_scoped_session
+
+    def __call__(self, task: Task, task_id, args, kwargs):
+        if task.request.retries > 0:
+            return
+        usr = kwargs.get("task_user_id")
+        if not usr:
+            return
+        now = datetime.datetime.now()
+        sa_session = self.ga_scoped_session
+        next_scheduled_time = self.calculate_task_start_time(usr, sa_session, self.task_exec_countdown_secs, now)
+        if next_scheduled_time > now:
+            count_down = next_scheduled_time - now
+            task.retry(countdown=count_down.total_seconds())
+
+    @abstractmethod
```

Comment:
I don't think you need this decorator since it's not a really an abstract method since it has a concrete implementation (unless you intended it as one, in which case it doesn't need to return `now`, I guess?)

Reply:
It is meant to be an abstract method. It returns now because one of the linters in the CI complained about it when it did not return anything. Other than the linter complaining the code worked fine when the abstract method did not return anything.